marconi-sidechain/src/Marconi/Sidechain/Api/HttpServer.hs

```diff
@@ -68,116 +69,101 @@ runHttpServer = do
       (marconiApp env)
 
 marconiApp :: SidechainEnv -> Application
-marconiApp env = serve (Proxy @API) (httpRpcServer env)
+marconiApp env =
+  serve (Proxy @API) $
+    hoistServer (Proxy @API) (hoistHandler env) httpRpcServer
 
-jsonRpcServer
-  :: SidechainEnv
-  -- ^  Utxo Environment to access Utxo Storage running on the marconi thread
-  -> Server JsonRpcAPI
-jsonRpcServer env =
+type ReaderHandler env = ExceptT ServerError (ReaderT env IO)
```

Comment:
Is there some function that's throwing a `ServerError`?

Reply:
Potentially any handler can. The standard way of returning a non-200 status is to use `throwError err400` or something similar.

The type `Handler` is actually a `newtype` of `ExceptT ServerError IO` and the hoisting just converts the `ReaderHandler` type here into that. It's changing the base monad from `IO` to `ReaderT env IO`, but to do that you have to unwrap the `ReaderHandler` down to `IO (Either ServerError a)` and then wrap it back up to `Handler` (ie `ExceptT ServerError IO`).

I haven't made `ReaderHandler` a `newtype` like `Handler` because I don't think that's needed when it's being used only within this file.

Reply:
FYI, `servant` doesn't have a good story around "real" (runtime) exceptions. They get converted into generic 500-status responses, which don't fit the schema for JSON-RPC errors (ie they're just text). I have a solution for this too, if you think it's worth pursuing.